vm/src/obj/objstr.rs

```diff
@@ -1072,6 +1090,200 @@ fn call_object_format(vm: &VirtualMachine, argument: PyObjectRef, format_spec: &
+                    match objint::get_value(&obj).to_u32().and_then(char::from_u32) {
+                        Some(value) => Ok(value.to_string()),
+                        None => {
+                            Err(vm.new_overflow_error("%c arg not in range(0x110000)".to_string()))
+                        }
+                    }
+                } else if objtype::isinstance(&obj, &vm.ctx.str_type()) {
+                    let s: String = get_value(&obj);
+                    let num_chars = s.chars().count();
+                    if num_chars != 1 {
+                        Err(vm.new_type_error("%c requires int or char".to_string()))
+                    } else {
+                        println!("Hurray!");
+                        Ok(s.chars().next().unwrap().to_string())
+                    }
+                } else {
+                    // TODO re-arrange this block so this error is only created once
+                    Err(vm.new_type_error("%c requires int or char".to_string()))
+                }
+            }?;
+            format_spec.precision = Some(CFormatQuantity::Amount(1));
+            Ok(format_spec.format_string(char_string))
+        }
+        _ => Err(vm.new_not_implemented_error(format!(
+            "Not yet implemented for %{}",
+            format_spec.format_char
+        ))),
+    }
+}
+
+fn do_cformat(
+    vm: &VirtualMachine,
+    mut format_string: CFormatString,
+    values_obj: PyObjectRef,
+) -> PyResult {
+    let mut final_string = String::new();
+    let num_specifiers = format_string
+        .format_parts
+        .iter()
+        .filter(|(_, part)| CFormatPart::is_specifier(part))
+        .count();
+    let mapping_required = format_string
+        .format_parts
+        .iter()
+        .any(|(_, part)| CFormatPart::has_key(part))
+        && format_string
+            .format_parts
+            .iter()
+            .filter(|(_, part)| CFormatPart::is_specifier(part))
+            .all(|(_, part)| CFormatPart::has_key(part));
+
+    let values_obj = if mapping_required {
+        if !objtype::isinstance(&values_obj, &vm.ctx.dict_type()) {
+            return Err(vm.new_type_error("format requires a mapping".to_string()));
+        }
+        values_obj.clone()
+    } else {
+        // check for only literal parts, in which case only empty tuple is allowed
+        if 0 == num_specifiers
+            && (!objtype::isinstance(&values_obj, &vm.ctx.tuple_type())
+                || !objtuple::get_value(&values_obj).is_empty())
+        {
+            return Err(vm.new_type_error(
+                "not all arguments converted during string formatting".to_string(),
+            ));
+        }
+
+        // convert `values_obj` to a new tuple if it's not a tuple
+        if !objtype::isinstance(&values_obj, &vm.ctx.tuple_type()) {
+            vm.ctx.new_tuple(vec![values_obj.clone()])
+        } else {
+            values_obj.clone()
+        }
+    };
+
+    let mut auto_index: usize = 0;
```

Comment:
Keeping this index seems somewhat weird to me. I would expect some sort of a `zip` operation over the format parts and the python objects to turn them into strings, then join all strings together. Off course it is not as simple as that since we have string literals in between, but maybe there is a way to store a sequence of pairs or string literals.

Reply:
Maybe creating a simple enum would do the trick:
```rust
enum StringPart {
  String,
  FormattedObject(FormatString, PyObject),
}
```

Reply:
> Keeping this index seems somewhat weird to me. I would expect some sort of a `zip` operation over the format parts and the python objects to turn them into strings, then join all strings together. Off course it is not as simple as that since we have string literals in between, but maybe there is a way to store a sequence of pairs or string literals.

It's indeed not as simple because the `values` object can be a mapping type, and not just a tuple.
I'll think of something tomorrow.

Reply:
@windelbouwman 
I don't have any elegant solution at the moment, the problems with zipping are
* `try_update_quantity_from_tuple` (asterik in format strings)
* `values_obj` might be a dict so it can't be converted to an iterator